src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs

```diff
@@ -251,80 +251,29 @@ private static int UnescapePercentEncoding(ref int scan, Span<byte> buffer)
             {
                 return -1;
             }
-
-            var probe = scan;
-
-            var value1 = ReadHex(ref probe, buffer);
-            if (value1 == -1)
+            
+            if (scan + 2 > buffer.Length)
```

Comment:
Possible integer overflow here. You should change this to be a subtraction on Length instead, or you end up having to do extra checks for overflow.

Also, the increment above has same issue.